opnode/rollup/driver/state.go

```diff
@@ -282,6 +313,31 @@ func (s *state) handleEpoch(ctx context.Context) (bool, error) {
 
 }
 
+func (s *state) handleUnsafeL2Payload(ctx context.Context, payload *l2.ExecutionPayload) error {
+	if s.l2SafeHead.Number > uint64(payload.BlockNumber) {
+		s.log.Info("ignoring unsafe L2 execution payload, already have safe payload", "id", payload.ID())
+		return nil
+	}
+
+	// Not that the payload may cause reorgs. The l2SafeHead may get out of sync because of this.
+	// The engine should never reorg past the finalized block hash however.
```

Comment:
We actually (correctly I believe), don't allow reorgs past the safe block head.

Reply:
The L1 safe head itself can still reorg, so even if L2 reorgs due to a block from p2p I don't think there's much of a difference here.

Reply:
In processBlock, we reject payloads behind the safe block head. You're right that the engine will not got reorg past the finalized hash, but this comment is still a little misleading about what exact reorgs it can cause.

Reply:
If you're referring to the remaining `s.l2SafeHead.Number > uint64(payload.BlockNumber)` heuristic in `handleUnsafeL2Payload`: a reorg past the safe block may still happen using a block newer than the safe-head if the common ancestor is before the safe head. So I think the comment is accurate that the `l2SafeHead` can get out of sync. The `processBlock` itself doesn't do any checks like that.

Reply:
You're right that is is on a different chain. I don't think that we should allow that reorg, but that might take more design to do.